api/types/resource_ids.go

```diff
@@ -52,18 +79,21 @@ func ResourceIDFromString(raw string) (ResourceID, error) {
 	if len(raw) < 1 || raw[0] != '/' {
 		return ResourceID{}, trace.BadParameter("%s is not a valid ResourceID string", raw)
 	}
-	raw = raw[1:]
-	// Should be safe for any Name as long as the ClusterName and Kind don't
-	// contain slashes, which should never happen.
-	parts := strings.SplitN(raw, "/", 3)
-	if len(parts) != 3 {
+	raw = strings.TrimRight(raw[1:], "/")
+	// Should be safe for any Name as long as the ClusterName, Kind and ResourceName
+	// don't contain slashes, which should never happen.
+	parts := strings.SplitN(raw, "/", 4)
```

Comment:
Hmm, that's bad. I will change the logic to accommodate those cases, but slashes in resource names aren't problematic for the UI?

Reply:
seems to work in the UI
<img width="143" alt="Screenshot 2023-01-10 at 2 34 00 PM" src="https://user-images.githubusercontent.com/16807901/211676725-08315422-465f-470a-9408-0014e40d11c3.png">

Reply:
I mean when using the API.

https://github.com/gravitational/teleport/blob/8b59b0d949012708598189561ac20e153b8a31ec/lib/web/apiserver.go#L572

Reply:
maybe some things won't work but we never explicitly block it, and ssh works from CLI and web UI, so I'm sure someone is using it

Reply:
The function is now changed to:
```go

// ResourceIDFromString parses a ResourceID from a string. The string should
// have been obtained from ResourceIDToString.
func ResourceIDFromString(raw string) (ResourceID, error) {
	if len(raw) < 1 || raw[0] != '/' {
		return ResourceID{}, trace.BadParameter("%s is not a valid ResourceID string", raw)
	}
	raw = raw[1:]
	// Should be safe for any Name as long as the ClusterName and Kind don't
	// contain slashes, which should never happen.
	parts := strings.SplitN(raw, "/", 3)
	if len(parts) != 3 {
		return ResourceID{}, trace.BadParameter("/%s is not a valid ResourceID string", raw)
	}
	resourceID := ResourceID{
		ClusterName: parts[0],
		Kind:        parts[1],
		Name:        parts[2],
	}
	if !slices.Contains(KubernetesResourcesKinds, resourceID.Kind) {
		return resourceID, trace.Wrap(resourceID.CheckAndSetDefaults())
	}
	// Kubernetes forbids slashes "/" in Namespaces and Pod names, so it's safe to
	// explode the resourceID.Name and extract the last two entries as namespace
	// and name.
	// Teleport allows the resource names to contain slashes, so we need to join
	// splits[:len(splits)-2] to reconstruct the resource name that contains slashes.
	// If splits slice does not have the correct size, resourceID.CheckAndSetDefaults()
	// will fail because, for kind=pod, it's mandatory to present a non-empty
	// namespace and name.
	splits := strings.Split(resourceID.Name, "/")
	if len(splits) >= 3 {
		resourceID.Name = strings.Join(splits[:len(splits)-2], "/")
		resourceID.SubResourceName = strings.Join(splits[len(splits)-2:], "/")
	}
	return resourceID, trace.Wrap(resourceID.CheckAndSetDefaults())
}


```